### ISSUE TITLE
Could not pass event VillagerAcquireTradeEvent to AdvancedTrades v1.0.8

### ISSUE THREAD

**Issue by Reldeam** [GC-spigot/AdvancedTrades#13]
### Describe the bug

When villagers are naturally generated in the game or when they level up I get this error and they don't gain any custom trades:

`18/01/2022 [10:38:42 ERROR: Could not pass event VillagerAcquireTradeEvent to AdvancedTrades v1.0.8]
java.lang.IllegalStateException: null
    at java.util.ArrayList$Itr.remove(ArrayList.java:980) ~[?:?]
    at net.advancedplugins.at.handlers.TradeBuilder.build(TradeBuilder.java:63) ~[AdvancedTrades-1.0.8.jar:?]
    at net.advancedplugins.at.listeners.VillagerTradesPopulator.onVillagerAcquireTradeEvent(VillagerTradesPopulator.java:38) ~[AdvancedTrades-1.0.8.jar:?]
    at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor86.execute(Unknown Source) ~[?:?]
    at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
    at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:git-Paper-66]
    at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
    at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
    at net.minecraft.world.entity.npc.AbstractVillager.addOffersFromItemListings(AbstractVillager.java:288) ~[?:?]
    at net.minecraft.world.entity.npc.Villager.updateTrades(Villager.java:951) ~[?:?]
    at net.minecraft.world.entity.npc.Villager.increaseMerchantCareer(Villager.java:781) ~[?:?]
    at net.minecraft.world.entity.npc.Villager.mobTick(Villager.java:258) ~[?:?]
    at net.minecraft.world.entity.npc.Villager.customServerAiStep(Villager.java:245) ~[?:?]
    at net.minecraft.world.entity.Mob.serverAiStep(Mob.java:865) ~[?:?]
    at net.minecraft.world.entity.LivingEntity.aiStep(LivingEntity.java:3251) ~[?:?]
    at net.minecraft.world.entity.Mob.aiStep(Mob.java:628) ~[?:?]
    at net.minecraft.world.entity.AgeableMob.aiStep(AgeableMob.java:154) ~[?:?]
    at net.minecraft.world.entity.LivingEntity.tick(LivingEntity.java:2907) ~[?:?]
    at net.minecraft.world.entity.Mob.tick(Mob.java:396) ~[?:?]
    at net.minecraft.world.entity.npc.Villager.tick(Villager.java:290) ~[?:?]
    at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1125) ~[?:?]
    at net.minecraft.world.level.Level.guardEntityTick(Level.java:970) ~[?:?]
    at net.minecraft.server.level.ServerLevel.lambda$tick$5(ServerLevel.java:667) ~[?:?]
    at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:43) ~[paper-1.18.jar:git-Paper-66]
    at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:647) ~[?:?]
    at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1607) ~[paper-1.18.jar:git-Paper-66]
    at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:480) ~[paper-1.18.jar:git-Paper-66]
    at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1470) ~[paper-1.18.jar:git-Paper-66]
    at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1264) ~[paper-1.18.jar:git-Paper-66]
    at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.18.jar:git-Paper-66]
    at java.lang.Thread.run(Thread.java:833) ~[?:?]`

### How to reproduce

1. Spawn villager that should have a custom trade possible.

### Server Log

https://mcpaste.io/c06b4818873db074

**Comment by Reldeam** [comment]
I've turned off all of my plugins except for AdvancedTrades and still got the error.

[log](https://mcpaste.io/45bbc415eb3b61e4)

My ***config.yml*** is default.

**trades.yml**
```
# Configuration file for Villager Trades
# Villagers Professions: https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/entity/Villager.Profession.html

trades:
  'sand': # Trade name, this can be anything you want
    tradeSettings:
      chance: 25
      requiredVillagerLevel: '2'
      requiredProfession: # Required villager profession, read more at https://trades.advancedplugins.net/trades/professions
        - 'Mason' # If set to none, required profession is disabled
      requiredBiome: # Required villager's biome for this trade, read more at https://trades.advancedplugins.net/trades/biomes
        - 'desert' # If set to none, required profession is disabled
      excludedWorlds: # What worlds should this trade not generate in?
        - 'none'
      maxTradeUses: '16'
    tradeIngredients: # Items required for a trade, read more at https://trades.advancedplugins.net/configuration/items
      # MAXIMUM OF 2 INGREDIENTS PER TRADE!
      '1':
        type: EMERALD
        amount: 1
    tradeResult: # Item which villager will offer in exchange of trade ingredients
      villager:
        experience: '1' # Experience gained by villager with this trade, read more at https://trades.advancedplugins.net/trades/levels
      item: # Item given by villager to player at trade, read more at https://trades.advancedplugins.net/configuration/items
        type: 'SAND'
        amount: 4
  'netherite_scrap': # Trade name, this can be anything you want
    tradeSettings:
      chance: 5
      requiredVillagerLevel: '5'
      requiredProfession: # Required villager profession, read more at https://trades.advancedplugins.net/trades/professions
        - 'Armorer' # If set to none, required profession is disabled
      requiredBiome: # Required villager's biome for this trade, read more at https://trades.advancedplugins.net/trades/biomes
        - 'nether_wastes'
        - 'soul_sand_valley'
        - 'crimson_forest'
        - 'warped_forest'
        - 'basalt_deltas'
      excludedWorlds: # What worlds should this trade not generate in?
        - 'none'
      maxTradeUses: '4'
    tradeIngredients: # Items required for a trade, read more at https://trades.advancedplugins.net/configuration/items
      # MAXIMUM OF 2 INGREDIENTS PER TRADE!
      '1':
        type: EMERALD
        amount: 20
      '2':
        type: NETHER_STAR
        amount: 1
    tradeResult: # Item which villager will offer in exchange of trade ingredients
      villager:
        experience: '5' # Experience gained by villager with this trade, read more at https://trades.advancedplugins.net/trades/levels
      item: # Item given by villager to player at trade, read more at https://trades.advancedplugins.net/configuration/items
        type: 'NETHERITE_SCRAP'
        amount: 2
```

**Comment by Reldeam** [comment]
I can replicate this with a base install of PaperMC 1.18 (build #66):

```
[08:49:49 INFO]: Environment: authHost='https://authserver.mojang.com', accountsHost='https://api.mojang.com', sessionHost='https://sessionserver.mojang.com', servicesHost='https://api.minecraftservices.com', name='PROD'
[08:49:49 INFO]: Reloading ResourceManager: Default, bukkit
[08:49:50 INFO]: Loaded 7 recipes
[08:49:50 INFO]: Starting minecraft server version 1.18
[08:49:50 INFO]: Loading properties
[08:49:50 INFO]: This server is running Paper version git-Paper-66 (MC: 1.18) (Implementing API version 1.18-R0.1-SNAPSHOT) (Git: dce79f3)
[08:49:51 INFO]: Server Ping Player Sample Count: 12
[08:49:51 INFO]: Using 4 threads for Netty based IO
[08:49:51 INFO]: Default game type: SURVIVAL
[08:49:51 INFO]: Generating keypair
[08:49:51 INFO]: Starting Minecraft server on *:25565
[08:49:51 INFO]: Using default channel type
[08:49:51 INFO]: Paper: Using Java 11 compression from Velocity.
[08:49:51 INFO]: Paper: Using Java cipher from Velocity.
[08:49:51 INFO]: Loading AdvancedTrades v1.0.8
[08:49:51 INFO]: Server permissions file permissions.yml is empty, ignoring it
[08:49:51 INFO]: Preparing level "world"
[08:49:51 INFO]: Preparing start region for dimension minecraft:overworld
[08:49:52 INFO]: Time elapsed: 122 ms
[08:49:52 INFO]: Preparing start region for dimension minecraft:the_nether
[08:49:52 INFO]: Time elapsed: 74 ms
[08:49:52 INFO]: Preparing start region for dimension minecraft:the_end
[08:49:52 INFO]: Time elapsed: 49 ms
[08:49:52 INFO]: Enabling AdvancedTrades v1.0.8
[08:49:53 INFO]: Successfully loaded 2 custom trades.
[08:49:53 INFO]: Running delayed init tasks
[08:49:53 INFO]: Done (2.281s)! For help, type "help"
[08:49:53 INFO]: Timings Reset
[08:51:28 INFO]: §d[AdvancedTrades]§r Reloading all configuration...
[08:51:28 INFO]: Successfully loaded 2 custom trades.
[08:51:28 INFO]: §d[AdvancedTrades]§r §aReloaded all configuration successfully
[08:57:08 INFO]: UUID of player Reldeam is 0691e8ce-fd57-4bd4-b328-b66277ab2d94
[08:57:08 INFO]: §eReldeam joined the game
[08:57:08 INFO]: Reldeam[/127.0.0.1:59979] logged in with entity id 226 at ([world]1.5, 67.0, 6.5)
[08:57:16 INFO]: Made Reldeam a server operator
[08:57:24 INFO]: Reldeam issued server command: /gamemode creative
[08:57:24 INFO]: §7§o[Reldeam: Set own game mode to Creative Mode]
[08:57:40 INFO]: Reldeam issued server command: /locatebiome minecraft:desert
[08:57:45 INFO]: Reldeam issued server command: /tp @s -3008 ~ -956
[08:57:45 INFO]: §7§o[Reldeam: Teleported Reldeam to -3007.500000, 71.045619, -955.500000]
[08:58:40 ERROR]: Could not pass event VillagerAcquireTradeEvent to AdvancedTrades v1.0.8
java.lang.IllegalStateException: null
	at java.util.ArrayList$Itr.remove(ArrayList.java:980) ~[?:?]
	at net.advancedplugins.at.handlers.TradeBuilder.build(TradeBuilder.java:63) ~[AdvancedTrades-1.0.8.jar:?]
	at net.advancedplugins.at.listeners.VillagerTradesPopulator.onVillagerAcquireTradeEvent(VillagerTradesPopulator.java:38) ~[AdvancedTrades-1.0.8.jar:?]
	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor2.execute(Unknown Source) ~[?:?]
	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:git-Paper-66]
	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:628) ~[paper-api-1.18-R0.1-SNAPSHOT.jar:?]
	at net.minecraft.world.entity.npc.AbstractVillager.addOffersFromItemListings(AbstractVillager.java:288) ~[?:?]
	at net.minecraft.world.entity.npc.Villager.updateTrades(Villager.java:951) ~[?:?]
	at net.minecraft.world.entity.npc.AbstractVillager.getOffers(AbstractVillager.java:129) ~[?:?]
	at net.minecraft.world.entity.ai.behavior.ShowTradesToPlayer.updateDisplayItems(ShowTradesToPlayer.java:122) ~[?:?]
	at net.minecraft.world.entity.ai.behavior.ShowTradesToPlayer.findItemsToDisplay(ShowTradesToPlayer.java:109) ~[?:?]
	at net.minecraft.world.entity.ai.behavior.ShowTradesToPlayer.tick(ShowTradesToPlayer.java:79) ~[?:?]
	at net.minecraft.world.entity.ai.behavior.ShowTradesToPlayer.canStillUse(ShowTradesToPlayer.java:21) ~[?:?]
	at net.minecraft.world.entity.ai.behavior.Behavior.tickOrStop(Behavior.java:75) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.world.entity.ai.Brain.tickEachRunningBehavior(Brain.java:526) ~[?:?]
	at net.minecraft.world.entity.ai.Brain.tick(Brain.java:471) ~[?:?]
	at net.minecraft.world.entity.npc.Villager.mobTick(Villager.java:248) ~[?:?]
	at net.minecraft.world.entity.npc.Villager.customServerAiStep(Villager.java:245) ~[?:?]
	at net.minecraft.world.entity.Mob.serverAiStep(Mob.java:865) ~[?:?]
	at net.minecraft.world.entity.LivingEntity.aiStep(LivingEntity.java:3251) ~[?:?]
	at net.minecraft.world.entity.Mob.aiStep(Mob.java:628) ~[?:?]
	at net.minecraft.world.entity.AgeableMob.aiStep(AgeableMob.java:154) ~[?:?]
	at net.minecraft.world.entity.LivingEntity.tick(LivingEntity.java:2907) ~[?:?]
	at net.minecraft.world.entity.Mob.tick(Mob.java:396) ~[?:?]
	at net.minecraft.world.entity.npc.Villager.tick(Villager.java:290) ~[?:?]
	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1125) ~[?:?]
	at net.minecraft.world.level.Level.guardEntityTick(Level.java:970) ~[?:?]
	at net.minecraft.server.level.ServerLevel.lambda$tick$5(ServerLevel.java:667) ~[?:?]
	at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:43) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:647) ~[?:?]
	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1607) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:480) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1470) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1264) ~[paper-1.18.jar:git-Paper-66]
	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:317) ~[paper-1.18.jar:git-Paper-66]
	at java.lang.Thread.run(Thread.java:833) ~[?:?]
```

**Comment by Reldeam** [comment]
So I decompiled your java file and had a quick look at the bytecode. It seems like when you are culling trade options in the _TradeBuilder_ class `TradeBuilder.build()` method, you are using a set of `if` statements instead of `else if` statements. The issue with that is if there are multiple conditions that do not match, then you try and remove the trade from the list of possible trades more than once:

```
public AdvancedTrade build() {
    // Byte code:
    //   0: getstatic net/advancedplugins/at/handlers/TradesHandler.a : Z
    //   3: new java/util/ArrayList
    //   6: dup
    //   7: invokestatic getTradesHandler : ()Lnet/advancedplugins/at/handlers/TradesHandler;
    //   10: invokevirtual getTrades : ()Lcom/google/common/collect/ImmutableCollection;
    //   13: invokespecial <init> : (Ljava/util/Collection;)V
    //   16: astore_2
    //   17: istore_1
    //   18: aload_2
    //   19: invokeinterface iterator : ()Ljava/util/Iterator;
    //   24: astore_3
    //   25: aload_3
    //   26: invokeinterface hasNext : ()Z
    //   31: ifeq -> 325
    //   34: aload_3
    //   35: invokeinterface next : ()Ljava/lang/Object;
    //   40: checkcast net/advancedplugins/at/trading/AdvancedTrade
    //   43: astore #4
    //   45: aload #4
    //   47: iload_1
    //   48: ifne -> 332
    //   51: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   54: invokevirtual getRequiredVillagerLevel : ()Ljava/lang/Integer;
    //   57: invokevirtual intValue : ()I
    //   60: iload_1
    //   61: ifne -> 126
    //   64: goto -> 68
    //   67: athrow
    //   68: ifle -> 111
    //   71: goto -> 75
    //   74: athrow
    //   75: aload #4
    //   77: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   80: invokevirtual getRequiredVillagerLevel : ()Ljava/lang/Integer;
    //   83: invokevirtual intValue : ()I
    //   86: goto -> 90
    //   89: athrow
    //   90: iload_1
    //   91: ifne -> 126
    //   94: aload_0
    //   95: getfield villagerLevel : I
    //   98: if_icmpeq -> 111
    //   101: aload_3
    //   102: invokeinterface remove : ()V
    //   107: goto -> 111
    //   110: athrow
    //   111: aload #4
    //   113: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   116: invokevirtual getRequiredBiome : ()Lcom/google/common/collect/ImmutableList;
    //   119: invokevirtual isEmpty : ()Z
    //   122: iload_1
    //   123: ifne -> 90
    //   126: iload_1
    //   127: ifne -> 197
    //   130: ifne -> 182
    //   133: goto -> 137
    //   136: athrow
    //   137: aload #4
    //   139: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   142: invokevirtual getRequiredBiome : ()Lcom/google/common/collect/ImmutableList;
    //   145: aload_0
    //   146: getfield biome : Lorg/bukkit/block/Biome;
    //   149: invokevirtual name : ()Ljava/lang/String;
    //   152: getstatic java/util/Locale.ROOT : Ljava/util/Locale;
    //   155: invokevirtual toLowerCase : (Ljava/util/Locale;)Ljava/lang/String;
    //   158: invokevirtual contains : (Ljava/lang/Object;)Z
    //   161: goto -> 165
    //   164: athrow
    //   165: iload_1
    //   166: ifne -> 197
    //   169: ifne -> 182
    //   172: aload_3
    //   173: invokeinterface remove : ()V
    //   178: goto -> 182
    //   181: athrow
    //   182: aload #4
    //   184: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   187: invokevirtual getExcludedWorlds : ()Lcom/google/common/collect/ImmutableList;
    //   190: invokevirtual isEmpty : ()Z
    //   193: iload_1
    //   194: ifne -> 165
    //   197: iload_1
    //   198: ifne -> 269
    //   201: ifne -> 250
    //   204: goto -> 208
    //   207: athrow
    //   208: aload #4
    //   210: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   213: invokevirtual getExcludedWorlds : ()Lcom/google/common/collect/ImmutableList;
    //   216: aload_0
    //   217: getfield world : Ljava/lang/String;
    //   220: getstatic java/util/Locale.ROOT : Ljava/util/Locale;
    //   223: invokevirtual toLowerCase : (Ljava/util/Locale;)Ljava/lang/String;
    //   226: invokevirtual contains : (Ljava/lang/Object;)Z
    //   229: goto -> 233
    //   232: athrow
    //   233: iload_1
    //   234: ifne -> 269
    //   237: ifeq -> 250
    //   240: aload_3
    //   241: invokeinterface remove : ()V
    //   246: goto -> 250
    //   249: athrow
    //   250: aload #4
    //   252: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   255: invokevirtual getRequiredProfession : ()Lcom/google/common/collect/ImmutableList;
    //   258: invokevirtual isEmpty : ()Z
    //   261: iload_1
    //   262: ifne -> 233
    //   265: iload_1
    //   266: ifne -> 308
    //   269: iload_1
    //   270: ifne -> 308
    //   273: ifne -> 321
    //   276: goto -> 280
    //   279: athrow
    //   280: aload #4
    //   282: invokevirtual getTradeSettings : ()Lnet/advancedplugins/at/trading/options/TradeSettings;
    //   285: invokevirtual getRequiredProfession : ()Lcom/google/common/collect/ImmutableList;
    //   288: aload_0
    //   289: getfield villagerProfession : Lorg/bukkit/entity/Villager$Profession;
    //   292: invokevirtual name : ()Ljava/lang/String;
    //   295: getstatic java/util/Locale.ROOT : Ljava/util/Locale;
    //   298: invokevirtual toLowerCase : (Ljava/util/Locale;)Ljava/lang/String;
    //   301: invokevirtual contains : (Ljava/lang/Object;)Z
    //   304: goto -> 308
    //   307: athrow
    //   308: ifne -> 321
    //   311: aload_3
    //   312: invokeinterface remove : ()V
    //   317: goto -> 321
    //   320: athrow
    //   321: iload_1
    //   322: ifeq -> 25
    //   325: invokestatic getTradesHandler : ()Lnet/advancedplugins/at/handlers/TradesHandler;
    //   328: aload_2
    //   329: invokevirtual getRandomTrade : (Ljava/util/Collection;)Lnet/advancedplugins/at/trading/AdvancedTrade;
    //   332: areturn
```

From what I can see here, it looks like the `goto` after each `remove` invocation is just to the next `if` statement.

So basically the end result is that if two more more conditions are not met on any of the trades, all trades fail. Please make the checks `else if` checks instead of a list of `if` checks to ensure the recipe is removed from the set of possible recipes only once.

**Comment by GC-spigot** [comment]
Fixed
